src/Gaffer/GraphComponent.cpp

```diff
@@ -318,14 +345,27 @@ void GraphComponent::removeChild( GraphComponentPtr child )
 
 	if( refCount() )
 	{
+		auto children = child->parent()->children();
```

Comment:
I present to you some interesting results :

```
import time

import Gaffer

s = Gaffer.ScriptNode()
for i in range( 0, 10000 ) :
	n = Gaffer.Node( "n{0}".format( i ) )
	s.addChild( n )
	s.selection().add( n )

c = time.clock()
s.deleteNodes( filter = s.selection() )
print time.clock() - c

# Master
# ------
#
# 0.06	0.06	0.05
#
# catalogueMultiSelect branch
# ---------------------------
#
# 1.95	1.91	1.92
```

We use a `vector<>` for storing children because it's the most compact memory-wise and in the most time-critical cases (hash/compute) we always access by index. But for large numbers of children it does open us up to dodgy quadratic scaling in cases like the above if we don't take great care. In [`ScriptNode::deleteNodes()`](https://github.com/GafferHQ/gaffer/blob/master/src/Gaffer/ScriptNode.cpp#L601) we deal with this by deleting the backmost nodes first, so I suspect searching from the back here is probably the answer too.

There's also the case where `addChild()` transfers a child from one parent to another and gets undone, which isn't dealt with here. And having some additional test cases is critical for something this fundamental. Unless you object, I'll take a stab at all this myself, so you can concentrate on the Catalogue stuff - sound OK?

Reply:
Sounds good. Thanks for taking care of it!

Reply:
So, it turns out that there's really not much overhead at all in the linear search to find the index. This was the actual problem :

```
auto children = child->parent()->children();
```

Even through `children()` returns `const ChildContainer &`, `auto` chooses `ChildContainer` and so takes a copy of the entire vector...